flow/diff_context.cc

```diff
@@ -128,23 +145,32 @@ Damage DiffContext::ComputeDamage(const SkIRect& accumulated_buffer_damage,
   return res;
 }
 
+SkRect DiffContext::MapRect(const SkRect& rect) {
+  SkRect mapped_rect;
+  if (!clip_tracker_.mapRect(rect, &mapped_rect)) {
+    // TODO(knopp): What to do here? We could possibly force repainting
+    // the whole frame?
+  }
+  return mapped_rect;
+}
+
 bool DiffContext::PushCullRect(const SkRect& clip) {
-  SkRect cull_rect = state_.transform.mapRect(clip);
-  return state_.cull_rect.intersect(cull_rect);
+  clip_tracker_.clipRect(clip, SkClipOp::kIntersect, false);
+  return !clip_tracker_.device_cull_rect().isEmpty();
```

Comment:
Maybe clip_tracker should have an `is_empty()` method. Especially since the cull_rect methods return a copy (device_cull_rect could potentially return a const reference, but that would be unwise since the referenced value may change if the clip_tracker is further operated upon).

Reply:
Note that clip_tracker has an is_empty for its stack, so it would have to be is_cull_rect_empty...

Reply:
Btw. the DisplayListMatrixClipTracker::Data::device_cull_rect() is virtual but not overriden anywhere. I'm pretty sure it gets devirtualized during LTO, but is it intentional?

Reply:
I think it was just virtual out of habit of that having been the interface for the 2 subclasses. How big of a deal is it to mark a method as virtual when it has no subclasses?